EIPS/eip-7377.md

```diff
@@ -0,0 +1,127 @@
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-07-21
+requires: 170, 1559, 2718
+---
+
+## Abstract
+
+Introduce a new [EIP-2718](./eip-2718.md) transaction type with the format `0x04 || rlp([chainId, nonce, maxFeePerGas, maxPriorityFeePerGas, gasLimit, codePtr, storage, data, value, accessList, yParity, r, s])` which sets the sending account's `code` field in the state trie to the `code` value at `codePtr` and applies the storage tuples to the sender's storage trie.
+
+## Motivation
+
+For as long as Ethereum has existed, so have complaints about the user experience of the platform.
+
+Ethereum was designed to be an abstract platform, limited only by the mind of the developers who deploy to it. Smart contract wallets were expected to placate unhappy users. Unfortunately, many years into this journey, the tension is still here.
+
+The ecosystem looks different than it did when when the first smart contract wallets were written. Several application-level standards have mitigated some of the issues. There are large standardization efforts around smart contract wallets, and a big push to make them the default on layer two chains. Everything seems to be moving in the right direction.
+
+One significant problems that hasn't been resolved though is what to do about the EOA users of *today*. Manually migrating assets from an EOA is not a tenable path to converting Ethereum's user base to smart contract wallets. We must provide a mechanism, embedded in the protocol, to migrate. This EIP proposes such mechanism.
+
+## Specification
+
+At the fork block `X`, introduce the MigrationTransaction transaction type.
+
+### MigrationTransaction
+
+#### Definition
+
+| field                  | type      |
+|------------------------|-----------|
+| `chainId`              | `int256`  |
+| `nonce`                | `uint64`  |
+| `maxFeePerGas`         | `int256`  |
+| `maxPriorityFeePerGas` | `int256`  |
+| `gasLimit`             | `uint64`  |
+| `codePtr`              | `address` |
+| `storage`              | `List[Tuple[uint256, uint256]]` |
+| `data`                 | `bytes`   |
+| `value`                | `int256`  |
+| `accessList`           | `List[Tuple[address, List[uint256]]]` |
+| `yParity`              | `uint8`   |
+| `v`                    | `uint256` |
+| `r`                    | `uint256` |
+
+The EIP-2718 `TransactionType` is `0x04` and the `TransactionPayload` is `rlp([chainId, nonce, maxFeePerGas, maxPriorityFeePerGas, gasLimit, codePtr, storageTuples, data, value, accessList, yParity, r, s])`.
+
+The transaction's signature hash is `keccak256(0x04 || rlp([chainId, nonce, maxFeePerGas, maxPriorityFeePerGas, gasLimit, codePtr, storageTuples, data, value, accessList])`
+
+#### Validation
+
+A MigrationTransaction is considered valid if the follow properties hold:
+
+* all [EIP-1559](./eip-1559.md) properties, unless specified otherwise
+* the code at `codePtr` is less than the [EIP-170](./eip-170) limit of `24576`
```

Comment:
```suggestion
* the code at `codePtr` is less than the [EIP-170](./eip-170.md) limit of `24576`
```
Nit: the extension is not optional.

CC @SamWilsn - this should be caught.

Reply:
gg